text/0000-trait-enums.md

```diff
@@ -0,0 +1,229 @@
+    }
+}
+```
+
+We must always make sure that the types in our enum implement the specified traits, otherwise the compiler will complain.
+
+```rust
+trait Animal {}
+
+struct Chicken;
+impl Animal for Chicken {}
+struct Sheep;
+
+// Error: Sheep does not implement trait Animal required by trait enum UnknownAnimal
+enum UnknownAnimal: Animal {
+    Chicken,
+    Sheep,
+}
+```
+
+This feature's usefulness might be hard to understand for new Rustaceans.
+However, more experienced Rust developpers know that using enums instead of trait objects is more optimized.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Trait enums are a middle ground between normal enums and C-like unions.
+They do have an active field like rust enums but the user does not need to know the contained type to use the trait enum.
+
+Trait enums must have at least one variant.
+
+## Initialization and assignment
+```rust
+enum Sword: Weapon {
+    Excalibur,
+    IronSword
+}
+
+let sword = Excalibur::new() as Sword; // Either like this
+let mut sword: Sword = IronSword::new(); // or like this
+sword = Excalibur::new(); // The type is already known so there's no need to specify it
+```
+
+## Downcasting
+Trait enums can be downcasted to their concrete type.
+
+```rust
+enum Cheese: Food {
+    BlueCheese,
+    GoatCheese
+}
+
+let cheese: Cheese = BlueCheese::new();
+if let Some(_) = cheese.downcast_ref::<GoatCheese>() {
+    println!("This is definitely Goat cheese!");
+}
+```
+
+## Conflicting methods in traits
+```rust
+trait Tool {
+    fn sharpen(&self);
+}
+
+trait Weapon {
+    fn sharpen(&self);
+}
+
+enum Sword: Tool + Weapon {
+    ...
+}
+
+let sword: Sword = IronSword::new();
+
+Tool::sharpen(&Sword);
+<Sword as Tool>::sharpen(&Sword);
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This feature is only useful in very specific cases and is quite complicated to implement.
+
+## Syntax
+This feature could either be named a trait enum, a type enum, a trait union or a safe union.
+Thus, it should either use the `enum` keyword or the `union` keyword.
+However, those keywords are already used so this would add more context-specific parsing.
+
+## Type inference
+The easiest way to easily manipulate trait enums would be to need a keyword like `Union` for unions.
+The keyword would be used to initialize and assign to trait enums.
+However, this is really ugly and hard to read. So it would be better if casting was needed or type inference didn't work at all.
+This makes the implementation more complicated.
+
+## Concrete type
+The concrete type is harder to get than with normal enums.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design allows achieving performance like normal enums while being convenient like trait objects.
+
+# Prior art
+[prior-art]: #prior-art
+
+This idea is brand new. There is no prior-art.
```

Comment:
Well, even if we ignore other languages, some relevant things from within the Rust community would be:
- past proposals on "sealed traits" like https://internals.rust-lang.org/t/closed-pre-rfc-sealed-traits-take-2/8857
- crates like [enum_derive](https://docs.rs/enum_derive/0.1.7/enum_derive/) that solve this for a handful of `std` traits
  - I strongly suspect a macro could also be written that takes any trait or method name and generates the boilerplate match block. There should definitely be an explanation why this doesn't work or isn't good enough, or else the whole RFC has no motivation.
- the current [delegation RFC](https://github.com/rust-lang/rfcs/pull/2393) which arguably tackles the same problem, but more generally than enums

Reply:
Come to think of it, a big chunk of delegation was "recently" implemented [as a procedural macro crate called "ambassador"](https://github.com/hobofan/ambassador). Maybe that could be extended to delegate single methods that aren't associated with a trait?

Reply:
Thanks a lot for your feedback! You are totally right, I should talk about why macros aren't good enough. I was not aware of those RFCs though so I will make the appropriate changes ASAP.

Reply:
Wow, the ambassador crate is awesome, I think there's no need to keep this open anymore!